bokeh/server/auth_provider.py

```diff
@@ -0,0 +1,318 @@
+        instead.
+
+        If a class is returned, it must be a subclass of RequestHandler,
+        which will used for the endpoint specified by ``logout_url``
+
+        '''
+        pass
+
+    @property
+    def login_url(self):
+        ''' A URL to redirect unathenticated users to for login.
+
+        This proprty may return None, if a ``get_login_url`` function is
+        supplied instead.
+
+        '''
+        pass
+
+    @property
+    def logout_hander(self):
+        ''' A request handler class for a logout page.
+
+        This property may return None.
+
+        If a class is returned, it must be a subclass of RequestHandler,
+        which will used for the endpoint specified by ``logout_url``
+
+        '''
+        pass
+
+    @property
+    def logout_url(self):
+        ''' A URL to redirect unathenticated users to for logout.
+
+        This proprty may return None.
+
+        '''
+        pass
+
+    def _validate(self):
+        if self.get_user and self.get_user_async:
+            raise ValueError("Only one of get_user or get_user_async should be supplied")
+
+        if (self.get_user or self.get_user_async) and not (self.login_url or self.get_login_url):
+            raise ValueError("When user authentication is enabled, one of login_url or get_login_url must be supplied")
+
+        if self.login_url and self.get_login_url:
+            raise ValueError("At most one of login_url or get_login_url should be supplied")
+        if self.login_handler and self.get_login_url:
+            raise ValueError("LoginHandler cannot be used with a get_login_url() function")
+        if self.login_handler and not issubclass(self.login_handler, RequestHandler):
+            raise ValueError("LoginHandler must be a Tornado RequestHandler")
+        if self.login_url and not probably_relative_url(self.login_url):
+            raise ValueError("LoginHandler can only be used with a relative login_url")
+
+        if self.logout_handler and not issubclass(self.logout_handler, RequestHandler):
+            raise ValueError("LogoutHandler must be a Tornado RequestHandler")
+        if self.logout_url and not probably_relative_url(self.logout_url):
+            raise ValueError("LogoutHandler can only be used with a relative login_url")
+
+class AuthModule(AuthProvider):
+    ''' An AuthProvider configured from a Python module.
+
+    The following properties return the corresponding values from the module if
+    they exist, or None otherwise:
+
+    * ``get_login_url``,
+    * ``get_user``
+    * ``get_user_async``
+    * ``login_url``
+    * ``logout_url``
+
+    The ``login_handler`` property will return a ``LoginHandler`` class from the
+    module, or None otherwise.
+
+    The ``logout_handler`` property will return a ``LogoutHandler`` class from
+    the module, or None otherwise.
+
+    '''
+
+    def __init__(self, module_path):
+        if not isfile(module_path):
+            raise ValueError()
```

Comment:
Maybe a message would make sense.

Reply:
I meant to go back and addd one but forgot, thank you

Reply:
https://github.com/bokeh/bokeh/pull/9190/commits/d06f2e2def0a503e11a51f6e372d8856ca5410b9